### PR TITLE
Add automated tests for executing tasks excluding Export, Upload, and Custom Task

### PR DESCRIPTION
**Pull request by svetogam**
More execution tests for #10:

* Expanded Copy Files
* Clear Directory Files
* Pack ZIP
* Sub-Routine
* Custom command in Bash
* Custom command in Windows

---

I made a few refactoring changes to the project outside of tests:

1. I changed referencing `Data.current_routine` to rely on it not being reset across execution, and then clearing it only in `Execution._exit_tree()`. This allows Execution to run without it being the highest-level scene (for sub-routines), since GutRunner is the highest-level scene during tests. It also makes things a little cleaner.
2. I emitted a `finished` signal in `Execution.finish()`. This helps to check against stalling in tests and to make tests faster and more reliable.

---

Notes on tests:

* The expanded Copy Files test found that copying a directory with Recursive=False does it recursively anyway, at least for me.
* I didn't actually run the Windows custom command test.
* Custom command tests rely on the working directory being the Project Builder project being tested. They won't work correctly otherwise. I think I can make this assumption, but if not then I don't think it would be worth keeping them around in this form.

### CODE REVIEW

**Comment by KoBeWi** [comment]
I ran the tests. 2 seem to be failing (`test_copy_files` and `test_windows_custom_task`), also in the latter there was an error due to `finish_mutext` being null.

In the windows custom taks, I'm getting `Could not create child process: copy Tests/Projects/TestProject1/File1.txt Tests/Projects/TestProject1/WindowsFile.txt` error. There are 2 problems here:
- CustomTask can't execute tasks that aren't specified by full path (on Windows at least); I think I ran into this myself recently. Needs more testing, might be engine bug.
- CustomTask does not get (pre)validated properly and fails at execution.

EDIT:
Well, CustomTask has no validation at all and not sure if it's feasible, as you can't reasonably check if a file exists in PATH (I think?). But the real problem is that this code does not account for `pipe_data` being empty, which can happen when execution fails:
https://github.com/KoBeWi/Godot-Project-Builds/blob/b87bc62a657c53be20ba9283a8c0b98ee623b496/Nodes/Command.gd#L48-L49
Also `copy` is not a program apparently, so it can't be executed as custom task. You can however run `cmd` with `/c copy`.

> I changed referencing Data.current_routine to rely on it not being reset across execution, and then clearing it only in Execution._exit_tree().

Yeah the previous "solution" was a really bad hack. I was thinking about reworking it too 😅 The idea behind that was to be 1000% sure that the current routine is reset properly.

> The expanded Copy Files test found that copying a directory with Recursive=False does it recursively anyway, at least for me.

https://github.com/KoBeWi/Godot-Project-Builds/blob/b87bc62a657c53be20ba9283a8c0b98ee623b496/Tasks/CopyFiles.tscn#L47-L51
This should be checking for `recursive.button_pressed` 🤦‍♂️
Feel free to fix that.

File: Tests/TestExecution.gd [hunk 1]
```diff
@@ -11,30 +26,240 @@ func before_all():
...
+	assert_true(files.has("SubDir/SubDirFile1.txt"))
+	
+	reader = ZIPReader.new()
+	error = reader.open(PROJECTS[1] + "Include.zip")
+	files = reader.get_files()
+	reader.close()
+	assert_true(error == OK)
+	assert_true(files.has("DirFile1.txt"))
+	assert_false(files.has("DirFile2.txt"))
+	assert_false(files.has("SubDir/SubDirFile1.txt"))
```
Comment (KoBeWi) [inline Tests/TestExecution.gd]: ZIP packing also allows for globbing (e.g. `*.txt`). Would be nice to test this too. (can be added later though)

**Comment by svetogam** [comment]
I removed the custom command tests. They look like a maintainance hassle and I'm not completely sure what I'm doing anyway.

I improved PackZIP tests to test globbing and some other things. It was weird to me that the filters work only per file and don't allow something like `SubDir/*`, but I'm guessing this is intentional? If not, I should change the Exclude test some more.

**Comment by KoBeWi** [comment]
> It was weird to me that the filters work only per file and don't allow something like SubDir/*, but I'm guessing this is intentional?

Well it's more or less intentional, I'm only matching the filename, not whole path. I think it's consistent with how Godot uses export filters.
Can always be changed if that's wanted of course.

**Comment by KoBeWi** [comment]
After pulling latest changes, the tests are failing for me
image

**Comment by svetogam** [comment]
It looks like some operations are slower in Godot v4.4-dev4. I think that's what caused the tests to fail.

I increased the waiting time for execution to finish from 5 to 20 seconds to be on the safe side. The longest test for me in 4.4-dev4 was 7.5 seconds.

It once happened to me that I had to wait the full 20 seconds even though execution apparently finished. That's annoying but I think it catches a bug that Execution.finish() is not always called when execution is done, so I guess it's fine to not hide that bug. I couldn't reproduce it.

You can modify the waiting time to whatever is convenient for you if 20 seconds is too long.

**Comment by KoBeWi** [comment]
> That's annoying but I think it catches a bug that Execution.finish() is not always called when execution is done, so I guess it's fine to not hide that bug.

This might be engine bug. I noticed the Project Builder hangs when it's supposed to finish a task.
The tests run fine in 4.3 (although they failed at my first attempt, idk what happened).

**Comment by KoBeWi** [comment]
Thanks!

**Comment by KoBeWi** [comment]
Actually the hang issue was caused by wrong use of threading. I previously made it work with a hack, but it stopped working. e2afa20c0e140c0e813f8b51102f06487c19856e makes it handled properly.

The one failed attempt test I mentioned was because I stopped the test mid-execution and it didn't do proper cleanup. Everything works fine.

**Pull request merged**